libplatsupport/src/plat/bcm2837/serial.c

```diff
@@ -53,6 +58,9 @@ int uart_putchar(ps_chardevice_t* d, int c)
 
 int uart_getchar(ps_chardevice_t* d UNUSED)
 {
+    if ((*REG_PTR(d->vaddr, MU_IIR) & MU_IIR_INTID_MASK) == MU_IIR_INTID_HOLD_EMPTY) {
+        return -1;
+    }
     while ( !(*REG_PTR(d->vaddr, MU_LSR) & MU_LSR_DATAREADY) );
```

Comment:
Now that we check `MU_IIR_INTID_HOLD_EMPTY, is the loop around `MU_LSR_DATAREADY` still necessary? Or is there an implicit guarantee that `MU_LSR_DATAREADY` is always set in this case, so we could remove this (or make it a comment).

Reply:
The manual is not really clear here. We could play safe and drop the loop, but still do exactly one check:
```
if (!(*REG_PTR(d->vaddr, MU_LSR) & MU_LSR_DATAREADY)) {
    /* this is not supposed to happen if MU_IIR_INTID says there is data. */
    return -1;
}